### DISCUSSION TITLE
Passing Client-Id and Client-Secret

### DISCUSSION THREAD

**Discussion by rasaliad** [viniciussanchez/RESTRequest4Delphi#99] [Q&A]
Hi Good day

How can I pass the Client-id and the Client-Secret to POST Request.

This is the post comand:

`procedure TFrmMain.btnPOSTClick(Sender: TObject);
var
  LResponse: IResponse;
begin
  LResponse := TRequest.New.BaseURL(edtBaseURL.Text)
    .Accept(edtAccept.Text)
    .AddBody(mmCustomBody.Lines.Text)
    .AddHeader('Client-id', edClientId.Text)
    .AddHeader('Client-secret', edClientSecret.Text)
    .Post;

  mmBody.Lines.Text := LResponse.Content;
  lblStatusCode.Caption := LResponse.StatusCode.ToString;
end;`

This is the Error a I get

`{"timestamp":"2022-08-03T12:20:23.098+00:00","status":400,"error":"Bad Request","message":"Missing request header 'client-id' for method parameter of type String","path":"/integration/hsusa/gadint"}
`
Thanks

**Comment by viniciussanchez** [comment]
Try .AddHeader('client-id', edClientId.Text)

**Reply by rasaliad** [reply]
Thanks Very much Mr. Sanchez

I change  `.AddHeader('Client-id', edClientId.Text)` for `.AddHeader('client-id', edClientId.Text)` and the error persist:

`{"timestamp":"2022-08-03T12:20:23.098+00:00","status":400,"error":"Bad Request","message":"Missing request header 'client-id' for method parameter of type String","path":"/integration/hsusa/gadint"} `

In Postman it is working, following is screenshot from postman:

![Screen Shot 2022-08-03 at 11 02 36 AM](https://user-images.githubusercontent.com/8439081/182646334-9348435a-eb87-4891-a7f0-989cfb0758df.png)
![Screen Shot 2022-08-03 at 11 01 03 AM](https://user-images.githubusercontent.com/8439081/182646363-17f191b8-500c-400c-89a7-a4944d5bf935.png)

What you think I'm doing wrong?

Thanks

**Comment by viniciussanchez** [comment]
.AddHeader('Client-id', edClientId.Text, [poDoNotEncode])

**Reply by rasaliad** [reply]
Hi Mr. Sanchez

Sorry for my ignorance, but I'm vey new Working with API.

But after I add the [poDoNotEncode] the get the following error:

`[dcc32 Error] Samples.Main.pas(85): E2034 Too many actual parameters`

I need to USES some another unit.

This is the event:

![Screen Shot 2022-08-03 at 2 28 41 PM](https://user-images.githubusercontent.com/8439081/182683538-db64e023-acb3-4da7-a22f-9a1b815584ec.png)


Thanks for your Help!!

**Reply by viniciussanchez** [reply]
Uses REST.Types
https://docwiki.embarcadero.com/Libraries/Sydney/en/REST.Types.TRESTRequestParameterOption

**Reply by rasaliad** [reply]
Ok, Mr.

Now the red underline disappears, but the .AddHeader('client-id', edClientId.Text, [poDoNotEncode]) only accept two parameters.

I'm using Delphi 10.2 Tokio.

`
uses RESTRequest4D,  REST.Types;

procedure TFrmMain.btnPOSTClick(Sender: TObject);
var
  LResponse: IResponse;
begin
  LResponse := TRequest.New.BaseURL(edtBaseURL.Text)
    .Accept(edtAccept.Text)
    .AddBody(mmCustomBody.Lines.Text)
    .AddHeader('client-id', edClientId.Text, [poDoNotEncode]) // the poDoNotEncode parameter is not accepted
    .AddHeader('client-secret', edClientSecret.Text, [poDoNotEncode]) // the poDoNotEncode parameter is not accepted    .Post;

  mmBody.Lines.Text := LResponse.Content;
  lblStatusCode.Caption := LResponse.StatusCode.ToString;
end;` 


Thanks,
